sql: introduce and use OptionalLiveness

tbg opened the pull request:
This makes it optional to provide a NodeLiveness instance.

The only uses are in DistSQL planning and jobs. In the former,
it was used only to check whether peers were live; we can just
avoid that check since no remote DistSQL planning occurs in
multi-tenancy Phase 2. For jobs, it was used to decide about job
adoption and cancellation, but again in multi-tenancy there is
at most one SQL server running (for that tenant) and so all jobs should
be adopted unconditionally.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48795)
<!-- Reviewable:end -->

tbg commented on pkg/jobs/registry.go:
```diff
@@ -1111,32 +1119,37 @@ WHERE status IN ($1, $2, $3, $4, $5) ORDER BY created DESC`
 		}
-		// Below we know that this node holds the lease on the job.
+
+		// Below we know that this node holds the lease on the job, or that we want
```
It's fine that we hit this with `runningOnNode == true`, right? It will hit `currentlyRunning` below and bail out, and also this is the same for single-tenant jobs.

ajwerner reviewed (commented):
Seems like there's a test that needs an update. Otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

cc @spaskob 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48795#-:-M7cJD5v7MPFFI9kQ3b3:b9bo8jp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @tbg)

___
*[pkg/jobs/helpers.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48795#-M7cGgsIAx4Yp498k4k2:-M7cGgsJ7xxQ3Iie6hqT:betbw64) ([raw file](https://github.com/cockroachdb/cockroach/blob/02cec22a29237fcceec16b26d1ed3e78c2ceebb2/pkg/jobs/helpers.go#L92)):*
> ```Go
> 
> func (nl *FakeNodeLiveness) IsLive(roachpb.NodeID) (bool, error) {
> 	return false, errors.New("unsupported here")
> ```

nit: use an error message that's a tad more specific and greppable 
___
*[pkg/jobs/registry.go, line 1151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48795#-M7DTev_74qnNr5sogFD-r1-1151:-M7cIuiVDGn6Ut34KrHo:b-u6fsfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9d17a7ed30710375ebd43ded16049f571362e4/pkg/jobs/registry.go#L1151)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It's fine that we hit this with `runningOnNode == true`, right? It will hit `currentlyRunning` below and bail out, and also this is the same for single-tenant jobs.
</blockquote></details>

Yes, that sounds right
___
*[pkg/jobs/registry.go, line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48795#-M7cGtAg4-JQadyIQeox:-M7cGtAg4-JQadyIQeoy:bkgtbhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/02cec22a29237fcceec16b26d1ed3e78c2ceebb2/pkg/jobs/registry.go#L538)):*
> ```Go
> 	nl, ok := nlw.Optional(47892)
> 	if !ok {
> 		// At most one container is running on behalf of a SQL tenant, so it must be
> ```

I hope we get that coordination right. Can we add a TODO here with my name on it to add a different leasing mechanism to not have to rely on this assumption? 


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=ajwerner

TFTR!

tbg commented:
bors r-

oopsed something

tbg commented:
bors r=ajwerner

tbg commented:
Fixed lint

bors r=ajwerner

tbg committed:
sql: introduce and use OptionalLiveness

This makes it optional to provide a NodeLiveness instance.

The only uses are in DistSQL planning and jobs. In the former,
it was used only to check whether peers were live; we can just
avoid that check since no remote DistSQL planning occurs in
multi-tenancy Phase 2. For jobs, it was used to decide about job
adoption and cancellation, but again in multi-tenancy there is
at most one SQL server running (for that tenant) and so all jobs should
be adopted unconditionally.

Release note: None

tbg commented:
Fixed lint again (staticcheck this time, not sure why that didn't pop up before)

bors r=ajwerner

tbg commented:
bors r=ajwerner
